src/Builder/BuilderEncoder.php

```diff
@@ -0,0 +1,293 @@
+    private function encodeAsSingle(OperatorInterface $value): stdClass
+    {
+        foreach (get_object_vars($value) as $val) {
+            $result = $this->recursiveEncode($val);
+
+            return $this->wrap($value, $result);
+        }
+
+        throw new LogicException(sprintf('Class "%s" does not have a single property.', $value::class));
+    }
+
+    private function encodeCombinedFilter(CombinedFieldQuery $filter): stdClass
+    {
+        $result = new stdClass();
+        foreach ($filter->fieldQueries as $filter) {
+            $filter = $this->recursiveEncode($filter);
+            if (is_object($filter)) {
+                $filter = get_object_vars($filter);
+            } elseif (! is_array($filter)) {
+                throw new LogicException(sprintf('Query filters must an array or an object. Got "%s"', get_debug_type($filter)));
+            }
+
+            foreach ($filter as $key => $value) {
+                $result->{$key} = $value;
+            }
+        }
+
+        return $result;
+    }
+
+    /**
+     * Query objects are encoded by merging query operator with field path to filter operators in the same object.
+     */
+    private function encodeQueryObject(QueryObject $query): stdClass
+    {
+        $result = new stdClass();
+        foreach ($query->queries as $key => $value) {
+            if ($value instanceof QueryInterface) {
+                // The sub-objects is merged into the main object, replacing duplicate keys
+                foreach (get_object_vars($this->recursiveEncode($value)) as $subKey => $subValue) {
+                    if (property_exists($result, $subKey)) {
+                        throw new LogicException(sprintf('Duplicate key "%s" in query object', $subKey));
+                    }
+
+                    $result->{$subKey} = $subValue;
+                }
+            } else {
+                if (property_exists($result, $key)) {
+                    throw new LogicException(sprintf('Duplicate key "%s" in query object', $key));
+                }
+
+                $result->{$key} = $this->encodeIfSupported($value);
+            }
+        }
+
+        return $result;
+    }
+
+    /**
+     * For the $setWindowFields stage output parameter, the optional window parameter is encoded in the same object
+     * of the window operator.
+     *
+     * @see https://www.mongodb.com/docs/manual/reference/operator/aggregation/setWindowFields/
+     */
+    private function encodeOutputWindow(OutputWindow $outputWindow): stdClass
+    {
+        $result = $this->recursiveEncode($outputWindow->operator);
+
+        // Transform the result into an stdClass if a document is provided
+        if (! $outputWindow->operator instanceof WindowInterface && (is_array($result) || is_object($result))) {
+            if (! is_first_key_operator($result)) {
+                throw new LogicException(sprintf('Expected OutputWindow::$operator to be an operator. Got "%s"', array_key_first($result)));
```

Comment:
[`array_key_first()`](https://www.php.net/array_key_first) requires an array parameter. This would fail if `$result` is an object (possible by the condition above).